### ISSUE TITLE
[BUG]: Error using Spark .NET Interactive Notebook Experiences on HDInsight Spark Cluster

### ISSUE THREAD

**Issue by fabiogouw** [dotnet/spark#413]
Hello. I'm trying these steps to enable .NET in Jupyter notebooks in a brand new HDInsight Spark Cluster:
https://github.com/dotnet/spark/tree/master/deployment/HDI-Spark/Notebooks

I'm able to see ".NET Spark (C#)" as an option while creating a new notebook. However, while submitting jobs using Spark .NET (for example, just typing _spark_ and sending this command), I'm receiving this error:

_The code failed because of a fatal error:
	Invalid status code '400' from http://headnodehost:8998/sessions with error payload: "Invalid kind: sparkdotnet (through reference chain: org.apache.livy.server.interactive.CreateInteractiveRequest[\"kind\"])"._

I also had to run these procedures https://github.com/MicrosoftDocs/azure-docs/blob/master/articles/hdinsight/hdinsight-release-notes.md because I was running the "ValueError: Cannot convert notebook to v5 because that version doesn't exist" issue.

Can you help me?

Thanks

**Comment by elvaliuliuliu** [comment]
Thanks for your question! Can you attach the full error and output logs from script action?

**Comment by fabiogouw** [comment]
Sure, here's the information requested:

```
Name
Install Sparkdotnet Interactive Notebook Experience
URI
https://myspark56hdistorage.blob.core.windows.net/scripts/install-interactive-notebook.sh
Parameters
0.6.0
Roles
Head,Worker
Status
Succeeded
Debug information
{
    "href": "http://hn0-myspar.bktuwdxhh4qe1nsazaxnbs2fta.nx.internal.cloudapp.net:8080/api/v1/clusters/myspark56/requests/31",
    "tasks": [
        {
            "href": "http://hn0-myspar.bktuwdxhh4qe1nsazaxnbs2fta.nx.internal.cloudapp.net:8080/api/v1/clusters/myspark56/requests/31/tasks/106",
            "Tasks": {
                "attempt_cnt": 1,
                "command": "ACTIONEXECUTE",
                "command_detail": "run_customscriptaction ACTIONEXECUTE",
                "end_time": 1580696084436,
                "error_log": "/var/lib/ambari-agent/data/errors-106.txt",
                "exit_code": 0,
                "host_name": "hn0-myspar.bktuwdxhh4qe1nsazaxnbs2fta.nx.internal.cloudapp.net",
                "id": "106",
                "output_log": "/var/lib/ambari-agent/data/output-106.txt",
                "request_id": "31",
                "role": "run_customscriptaction",
                "stage_id": "0",
                "start_time": 1580696067891,
                "status": "COMPLETED",
                "stderr": null,
                "stdout": null,
                "structured_out": null
            }
        },
        {
            "href": "http://hn0-myspar.bktuwdxhh4qe1nsazaxnbs2fta.nx.internal.cloudapp.net:8080/api/v1/clusters/myspark56/requests/31/tasks/107",
            "Tasks": {
                "attempt_cnt": 1,
                "command": "ACTIONEXECUTE",
                "command_detail": "run_customscriptaction ACTIONEXECUTE",
                "end_time": 1580696089526,
                "error_log": "/var/lib/ambari-agent/data/errors-107.txt",
                "exit_code": 0,
                "host_name": "hn1-myspar.bktuwdxhh4qe1nsazaxnbs2fta.nx.internal.cloudapp.net",
                "id": "107",
                "output_log": "/var/lib/ambari-agent/data/output-107.txt",
                "request_id": "31",
                "role": "run_customscriptaction",
                "stage_id": "0",
                "start_time": 1580696067891,
                "status": "COMPLETED",
                "stderr": null,
                "stdout": null,
                "structured_out": null
            }
        },
        {
            "href": "http://hn0-myspar.bktuwdxhh4qe1nsazaxnbs2fta.nx.internal.cloudapp.net:8080/api/v1/clusters/myspark56/requests/31/tasks/108",
            "Tasks": {
                "attempt_cnt": 1,
                "command": "ACTIONEXECUTE",
                "command_detail": "run_customscriptaction ACTIONEXECUTE",
                "end_time": 1580696167372,
                "error_log": "/var/lib/ambari-agent/data/errors-108.txt",
                "exit_code": 0,
                "host_name": "wn1-myspar.bktuwdxhh4qe1nsazaxnbs2fta.nx.internal.cloudapp.net",
                "id": "108",
                "output_log": "/var/lib/ambari-agent/data/output-108.txt",
                "request_id": "31",
                "role": "run_customscriptaction",
                "stage_id": "0",
                "start_time": 1580696067891,
                "status": "COMPLETED",
                "stderr": null,
                "stdout": null,
                "structured_out": null
            }
        },
        {
            "href": "http://hn0-myspar.bktuwdxhh4qe1nsazaxnbs2fta.nx.internal.cloudapp.net:8080/api/v1/clusters/myspark56/requests/31/tasks/109",
            "Tasks": {
                "attempt_cnt": 1,
                "command": "ACTIONEXECUTE",
                "command_detail": "run_customscriptaction ACTIONEXECUTE",
                "end_time": 1580696157074,
                "error_log": "/var/lib/ambari-agent/data/errors-109.txt",
                "exit_code": 0,
                "host_name": "wn2-myspar.bktuwdxhh4qe1nsazaxnbs2fta.nx.internal.cloudapp.net",
                "id": "109",
                "output_log": "/var/lib/ambari-agent/data/output-109.txt",
                "request_id": "31",
                "role": "run_customscriptaction",
                "stage_id": "0",
                "start_time": 1580696067891,
                "status": "COMPLETED",
                "stderr": null,
                "stdout": null,
                "structured_out": null
            }
        },
        {
            "href": "http://hn0-myspar.bktuwdxhh4qe1nsazaxnbs2fta.nx.internal.cloudapp.net:8080/api/v1/clusters/myspark56/requests/31/tasks/110",
            "Tasks": {
                "attempt_cnt": 1,
                "command": "ACTIONEXECUTE",
                "command_detail": "run_customscriptaction ACTIONEXECUTE",
                "end_time": 1580696166797,
                "error_log": "/var/lib/ambari-agent/data/errors-110.txt",
                "exit_code": 0,
                "host_name": "wn4-myspar.bktuwdxhh4qe1nsazaxnbs2fta.nx.internal.cloudapp.net",
                "id": "110",
                "output_log": "/var/lib/ambari-agent/data/output-110.txt",
                "request_id": "31",
                "role": "run_customscriptaction",
                "stage_id": "0",
                "start_time": 1580696067891,
                "status": "COMPLETED",
                "stderr": null,
                "stdout": null,
                "structured_out": null
            }
        },
        {
            "href": "http://hn0-myspar.bktuwdxhh4qe1nsazaxnbs2fta.nx.internal.cloudapp.net:8080/api/v1/clusters/myspark56/requests/31/tasks/111",
            "Tasks": {
                "attempt_cnt": 1,
                "command": "ACTIONEXECUTE",
                "command_detail": "run_customscriptaction ACTIONEXECUTE",
                "end_time": 1580696161926,
                "error_log": "/var/lib/ambari-agent/data/errors-111.txt",
                "exit_code": 0,
                "host_name": "wn5-myspar.bktuwdxhh4qe1nsazaxnbs2fta.nx.internal.cloudapp.net",
                "id": "111",
                "output_log": "/var/lib/ambari-agent/data/output-111.txt",
                "request_id": "31",
                "role": "run_customscriptaction",
                "stage_id": "0",
                "start_time": 1580696067891,
                "status": "COMPLETED",
                "stderr": null,
                "stdout": null,
                "structured_out": null
            }
        }
    ]
}
```
And the error
```
The code failed because of a fatal error:
	Invalid status code '400' from http://headnodehost:8998/sessions with error payload: "Invalid kind: sparkdotnet (through reference chain: org.apache.livy.server.interactive.CreateInteractiveRequest[\"kind\"])".

Some things to try:
a) Make sure Spark has enough available resources for Jupyter to create a Spark context. For instructions on how to assign resources see http://go.microsoft.com/fwlink/?LinkId=717038
b) Contact your cluster administrator to make sure the Spark magics library is configured correctly.
```
The lines I've tried were
```
int i = 0;
Command finishedat 02-02-2020 23:22:51.133 -03:00, execution took 147ms
```
and
```
spark
Command finishedat 02-02-2020 23:23:03.352 -03:00, execution took 3ms
```
Thank you

**Comment by imback82** [comment]
```
"Invalid kind: sparkdotnet (through reference chain: org.apache.livy.server.interactive.CreateInteractiveRequest["kind"])".
``` 
This means that the new livy jars are not being picked up. Most likely you need to remove (please back up) existing livy jars. They should be located under `/usr/hdp/current/livy2-server/`.

I see the following in the installation script:
```
sudo wget https://sparkdotnetrepl.blob.core.windows.net/notebooks/livy_jar.zip
sudo unzip -o livy_jar.zip
sudo cp -rf livy_jar/* /usr/hdp/current/livy2-server/
```

I would start looking at the `livy_jar.zip` contents and remove the same files (version should be different) from `/usr/hdp/current/livy2-server/`.

@elvaliuliuliu I think we need to update the script to handle this case.

**Comment by elvaliuliuliu** [comment]
@imback82 Got it, there might be some version different in livy jars, even new livy jars have been copied over, old livy jars (different version) still exist so that new jars are not being picked up. I will update the script to handle this case.

@fabiogouw Have you tried @imback82 's suggestion? Does that work for you?

**Comment by fabiogouw** [comment]
@elvaliuliuliu, after putting the following lines in the script and running it again I was able to use .NET in Jupyter! Thank you and @imback82 !

```
sudo rm /usr/hdp/current/livy2-server/jars/*.jar
sudo rm /usr/hdp/current/livy2-server/rsc-jars/*.jar
sudo rm /usr/hdp/current/livy2-server/repl_2.11-jars/*.jar
```
Ps. No problem just deleting the jars for me, it's just a sandbox cluster.

And about the contents of that folder
```
sshuser@hn0-toobig:/usr/hdp/current/livy2-server$ ls
bin  conf  DISCLAIMER  etc  jars  LICENSE  logs  NOTICE  repl_2.11-jars  rsc-jars
```

**Comment by elvaliuliuliu** [comment]
@fabiogouw, Thanks for your updates! This case should be fixed by #415.

**Comment by imback82** [comment]
Fixed by #415
